### ISSUE TITLE
Support for Yarn package manager

### ISSUE THREAD

**Issue by amasad** [heroku/heroku-buildpack-nodejs#337]
Yarn is a new package manager -- an npm replacement -- by Facebook and Google. It's faster, more reliable and secure than npm. I'd like to use it to manage my Heroku node app dependencies.

**Comment by hunterloftis** [comment]
So would I. This feature will be coming soon.

**Comment by rstacruz** [comment]
:+1:

If I may suggest, maybe an ENV config variable like `USE_YARN=1` would do.

**Comment by kane-thornwyrd** [comment]
> It's faster, more reliable and secure than npm

Fact please.

**Comment by amasad** [comment]
@kane-thornwyrd the website has tons of facts for you: https://yarnpkg.com/

**Comment by rstacruz** [comment]
Oh, and if I may suggest again, the presence of a yarn.lock file should automatically use yarn for you. 😊

**Comment by intellix** [comment]
@kane-thornwyrd Performance: https://yarnpkg.com/en/compare
Reliable: The results are deterministic. The amount of times I've had to tell members of my team to: "it worked on my machine... try again maybe? great, it worked this time"

**Comment by kane-thornwyrd** [comment]
Sorry everyone if it may sound like defiance, I think that's a good thing to add Yarn to Heroku, and anyway my opinion does not count about this, and sorry to clutter this demand, I just seek some rock solid facts about yarn among its early supporters.

@intellix thank you for your answers, I have more indepth concerns:

**Performances** Without a proper protocol, checking which servers are requested, the network quality during the tests and some other data, I'm not particullary receptive to this kind of graph. 
Moreover, _in my case_, given I don't consider the re-/installation like something done often, perfs here are a margin interest. But, okay, that's a cool thing to have anyway.

**Reliable** In 3 years the only package manager related issue I witnessed, among my heteroclite teams (Ubuntu, Fedora, Win7, Win10,Debian over SSH on Win7, OS X...), was all the binary dependencies requiring node-gyp/Visual Studio/XCode/etc.
What you're decribing is quite strange, I mean, you should have worked on ultra complex app with dependencies on multiple versions of the same libraries, and in this case the _Flat mode_ which:
> Resolve mismatching versions of dependencies to a single version to avoid creating duplicates

Would ruin your day, am I wrong ?

**Comment by technomage** [comment]
More secure because it records checksums of installed packages on update and verifies on install. This prevents a man in the middle from substituting a rogue package.

**Comment by technomage** [comment]
Typical non version issues result from loose package.json (*, >=, ^) that result in different versions when installed at different times. In teams that religiously npm install after each pull and clear out old versions, this may not be an issue.

**Comment by kane-thornwyrd** [comment]
@technomage : thank you for your answer, I don't get how yarn is supposed to authenticate packages since it is supposed to support custom repos, then I guess it use the sha-1 digest to ensure a proper download, as npm. That does not prevent a MitM attack. Does it store the hashes ? Then how it is supposed to be aware of the new ones from a minor release ?
It is some serious innovation in terms of security if it can certify a link in the chain of trust without controlling the previous !

**Comment by hunterloftis** [comment]
This discussion is getting into the weeds. We at Heroku, and more importantly our users, feel the same pains with npm as described in Facebook's announcement page. We're very excited about Yarn and will be adding support for it (I'm working on that right at this moment). Everything has tradeoffs, but there are better places to discuss them :)

**Comment by hunterloftis** [comment]
Give it a shot:

```
heroku buildpacks:set https://github.com/heroku/heroku-buildpack-nodejs#yarn
```

**Comment by lfittl** [comment]
@hunterloftis Works well here - thanks for the super-fast turnaround on this :)

**Comment by parse** [comment]
@hunterloftis The `postinstall` script doesnt seem to be executed on Heroku using the build pack above. It looks like it is fixed in https://github.com/yarnpkg/yarn/pull/800, do you think its related?

**Comment by drewhamlett** [comment]
Only fact needed is to know Yehuda Katz (Bundler, Cargo) is involved

**Comment by joshmanders** [comment]
Any status on this being merged in and a new version cut so that I can put it into my build system?

**Comment by amasad** [comment]
@hunterloftis this works well! Only issue, as @parse mentioned this doesn't run `postinstall`. However, I was able to get around this by using `heroku-postbuild` custom script hook.

**Comment by voldern** [comment]
> @hunterloftis this works well! Only issue, as @parse mentioned this doesn't run postinstall. 

This has been fixed in v0.16.0 that was released a day ago. See https://github.com/yarnpkg/yarn/pull/800

**Comment by steffenagger** [comment]
Yep, this works @hunterloftis. Maybe include a version check looking at `engines.yarn`?

**Comment by mhuebert** [comment]
@hunterloftis thanks for getting this out so fast!

I've discovered a problem related to caching, and pushed a minimal example with steps to reproduce the issue: https://github.com/mhuebert/heroku-yarn-test

The gist: if a project depends on "bcrypt", the first build/heroku-push will work, but subsequent builds (after changing dependencies) will fail with an error:

```
/app/node_modules/bindings/bindings.js:91
  ^

Error: Could not locate the bindings file. Tried:
 → /app/node_modules/bcrypt/build/Debug/bcrypt_lib.node
 → /app/node_modules/bcrypt/build/bcrypt_lib.node
 → /app/node_modules/bcrypt/build/Release/bcrypt_lib.node
 → /app/node_modules/bcrypt/out/Debug/bcrypt_lib.node
 → /app/node_modules/bcrypt/Debug/bcrypt_lib.node
```

After this happens, the only way I can get any subsequent build to work is by using [heroku-repo](https://github.com/heroku/heroku-repo#purge_cache) to purge the cache (`heroku repo:purge_cache`)

**Comment by ChiperSoft** [comment]
I've been running into the bcrypt problem as well on my codeship builds, so I don't think it's a heroku problem. Something is wrong with the way yarn replaces the bindings when it re-runs install scripts. I had to start deleting my node_modules folder before every build.

**Comment by hunterloftis** [comment]
This will disable the cache (for modules like bcrypt that don't yet work well with yarn's cache):

```
$ heroku config:set NODE_MODULES_CACHE=false
```

**Comment by tillkruss** [comment]
I've set:

```
heroku buildpacks:set https://github.com/heroku/heroku-buildpack-nodejs#yarn
```

But the deploy fails. Any tips?

```
-----> Node.js app detected
-----> Creating runtime environment
       
       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_ENV=production
       NODE_MODULES_CACHE=true
-----> Installing binaries
       engines.node (package.json):  unspecified
       engines.npm (package.json):   unspecified (use default)
       
       Resolving node version (latest stable) via semver.io...
       Downloading and installing node 5.11.1...
       Using default npm version: 3.8.6
-----> Restoring cache
       Loading 2 from cacheDirectories (default):
       - node_modules
       - bower_components (not cached - skipping)
-----> Building dependencies
       Installing node modules (package.json)
-----> Caching build
       Clearing previous node cache
       Saving 2 cacheDirectories (default):
       - node_modules
       - bower_components (nothing to cache)
-----> Build succeeded!
       └── (empty)
       
-----> PHP app detected
-----> Bootstrapping...
-----> Installing platform packages...
[...snip...]
-----> Installing dependencies...
       Composer version 1.2.1 2016-09-12 11:27:19
       Loading composer repositories with package information
       Installing dependencies from lock file
[...snip...]
-----> Running 'composer compile'...
       > (cd web/app/themes/caroline && yarn)
       sh: 1: yarn: not found
       Script (cd web/app/themes/caroline && yarn) handling the compile event returned with error code 127
 !     Push rejected, failed to compile PHP app.
 !     Push failed
```

**Comment by natecox** [comment]
@kane-thornwyrd here are practical CI benchmarks for yarn, if these haven't already been linked somewhere above: https://medium.freecodecamp.com/npm-vs-yarn-benchmark-9b456de4aa96#.sa32q3heg

**Comment by joneslee85** [comment]
@kane-thornwyrd I ran into more issues with `npm` such as sometimes the npm failed to fetch, errors out with 404. Plus `npm` runs slow as hell. That alone gives enough push for `yarn` in Heroku. I am not jumping on the hype bandwagon, but I have to admit `npm` does not deliver the reliability that users deserve.

**Comment by kane-thornwyrd** [comment]
@joneslee85 as @hunterloftis said, there is other places to discuss this topic.

**Comment by hunterloftis** [comment]
@tillkruss looks like perhaps the node buildpack isn't exporting `yarn` properly for subsequent buildpacks in a multi-buildpack setup. I'll take a look.

@joshmanders merge into master will take a little while; needs a refactor (the support for yarn right now is ugly / hard to test / hard to maintain).

**Comment by tillkruss** [comment]
@hunterloftis: Thanks!

**Comment by robinbullocks4rb** [comment]
@hunterloftis just tried out the buildpack on a test app and it makes some progress before falling over when trying to install a package from a GitHub tarball (https://github.com/robinbullocks4rb/exceljs/tarball/replace-unzip2). It does work fine in my local environment.

Not sure if you would even expect this to work at this stage, but wanted to put it out there. Happy to raise a separate issue if it's easier.

```
remote:        [1/4] Resolving packages...
remote:        [2/4] Fetching packages...
remote:        error https://github.com/robinbullocks4rb/exceljs/tarball/replace-unzip2: Bad hash. Expected 600fb915da2cc2633d3ae82817494e92c79731eb but got 6bb5ae50dbdfa82af0ffe6cb7bfb77b2aaed8fe9
remote:        info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Interestingly, we have other tarball depdencies and they don't fail in the same way (e.g. if I remove the aforementioned one). However, we just get a more general failure message. I'm not sure how I would inspect the `yarn-error.log` for more info.

```
remote:        [1/4] Resolving packages...
remote:        [2/4] Fetching packages...
remote:        error An unexpected error occured, please open a bug report with the information provided in "/tmp/build_ab87074fd2ae4187c5576d13c9645bb9/yarn-error.log".
remote:        info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Finally, I tried `heroku run bash` on an existing app and tried running `yarn` there and the installation was successful, so the problem would appear to be buildpack-specific.

PS Thanks for addressing Yarn on Heroku so fast!

**Comment by tillkruss** [comment]
@hunterloftis I added an empty `yarn.lock` file at the project root and it works. My real `yarn.lock` files is further down the directory tree.

**Comment by janpaul123** [comment]
We should also definitely use [pure-lockfile](https://yarnpkg.com/en/docs/cli/install#toc-yarn-install-pure-lockfile), otherwise unexpected versions may be installed.

**Comment by RochesterinNYC** [comment]
@hunterloftis noting that there's [demand for this in the Cloud Foundry nodejs buildpack as well](https://github.com/cloudfoundry/nodejs-buildpack/pull/71). Is there any way we can help?

**Comment by kauffecup** [comment]
In the current state of the Yarn branch it still uses npm when doing a rebuild. To update rebuilds to also use Yarn, I've opened https://github.com/heroku/heroku-buildpack-nodejs/pull/341

**Comment by amasad** [comment]
This started failing this morning (it's been working fine for weeks) at the step where it installs yarn:

```
remote:        Downloading and installing yarn...
remote: /tmp/build_fa7fc117748f244d57f5673898c9a15d/.heroku/yarn/bin/yarn.js:47
remote:       throw err;
remote:       ^
remote: 
remote: Error: ENOENT: no such file or directory, open '/app/.cache/yarn/.roadrunner.json'
remote:     at Error (native)
remote:     at Object.fs.openSync (fs.js:549:18)
remote:     at Object.fs.writeFileSync (fs.js:1156:15)
remote:     at /tmp/build_fa7fc117748f244d57f5673898c9a15d/.heroku/yarn/node_modules/roadrunner/index.js:25:6
remote:     at /tmp/build_fa7fc117748f244d57f5673898c9a15d/.heroku/yarn/node_modules/roadrunner/index.js:12:12
remote:     at emitOne (events.js:82:20)
remote:     at process.emit (events.js:169:7)
remote:     at processEmit [as emit] (/tmp/build_fa7fc117748f244d57f5673898c9a15d/.heroku/yarn/node_modules/signal-exit/index.js:140:35)
remote:     at process.exit (node.js:766:17)
remote:     at Command.<anonymous> (/tmp/build_fa7fc117748f244d57f5673898c9a15d/.heroku/yarn/node_modules/commander/index.js:825:13)
```

**Comment by pasupulaphani** [comment]
@heroku-mirror can we have yarn buildpacks tagged with versions?

**Comment by Gpx** [comment]
FYI Yarn fixed the issues in 0.17.4.

I really like the approach suggested by @pasupulaphani

**Comment by joeyespo** [comment]
Looks like this early Yarn release doesn't work with `NPM_CONFIG_PRODUCTION=false` ([as described here](https://devcenter.heroku.com/articles/nodejs-support#devdependencies)).

Everything works otherwise when manually setting `NODE_ENV=development` in `app.json`. (Nice work, everyone!) Does anyone know if there's an equivalent ENV for Yarn to install devDependencies without relying on `NODE_ENV`?

**Comment by joeyespo** [comment]
Followup:

Looks like [this issue](https://github.com/yarnpkg/yarn/pull/1292) might be related (and [this comment](https://github.com/yarnpkg/yarn/pull/1292#issuecomment-256604256)). Looks like Yarn *does* take `NPM_CONFIG_*` vars into account, but `'false'` was being treated as `true` since it's a non-empty string.

Also, for reference, Yarn's `NPM_CONFIG_*` equivalent is `YARN_*`.

**Comment by joeyespo** [comment]
**Followup 2**

I still see the problem on the latest Yarn release.

The real issue seems to be when both `NODE_ENV=production` and `NPM_CONFIG_PRODUCTION=false` are set. `npm` installs devDependencies in this case and yarn doesn't.

So this is a yarn bug. I just [opened a new issue there](https://github.com/yarnpkg/yarn/issues/1975). Thanks!

**Comment by b0gok** [comment]
How can I setup yarn with `--ignore-engines` option?

**Comment by joeyespo** [comment]
**Update**

[A fix](https://github.com/yarnpkg/yarn/issues/2057) for [that](https://github.com/heroku/heroku-buildpack-nodejs/issues/337#issuecomment-262101793) issue was just merged into Yarn. Once Yarn releases a new version (and their latest regression from #343 is addressed), the `NPM_CONFIG_PRODUCTION=false` case should work.

**Comment by ignatiusreza** [comment]
cross reporting with https://github.com/yarnpkg/yarn/issues/785

got issue with missing modules possibly caused by yarn failing to install sub-dependencies.. though, weirdly, installation on local environment was working fine, it failed only when deploying to heroku using this buildpack.. the workaround mentioned in yarn issue (to do `yarn add [package]`) fix the deploy..

**Comment by joshmanders** [comment]
@ignatiusreza are you using anything that has fsevents as a dependency? Seems that is the root of most of the issues, I get it on all my stuff because I use `node-sass` which has it as a dependency.

**Comment by dylanjha** [comment]
I also had a module missing, but it wasn't the same as you guys:

> Cannot find module 'rechoir'

rechoir is a dependency of `liftoff@^2.1.0` which is a dependency of `gulp@3.9.0` and it was not installed.

**Comment by ignatiusreza** [comment]
@joshmanders I do have package that has fsevents as optional dependencies, the dependency tree goes like:

```
├─┬ webpack@1.14.0
  ├─┬ watchpack@0.2.9
    └─┬ chokidar@1.6.1 (optionally depends on fsevents)
```
though the ones missing in my case were `lodash._isiterateecall`, `pkg-dir`, `sax`, `has`..

and the dependencies goes like:

```
├─┬ assets-webpack-plugin@3.5.0
│ ├─┬ lodash.assign@3.2.0
│   ├─┬ lodash._createassigner@3.1.1
│     ├── lodash._isiterateecall@3.0.9 (missing)
├─┬ babel-loader@6.2.9
│ ├─┬ find-cache-dir@0.1.1
│   └─┬ pkg-dir@1.0.0 (missing)
├─┬ css-loader@0.25.0
│ ├─┬ cssnano@3.9.1
│   ├─┬ has@1.0.1 (missing)
│   ├─┬ postcss-svgo@2.1.5
│     └─┬ svgo@0.7.1
│       ├── sax@1.2.1 (missing)
```

**Comment by hunterloftis** [comment]
Resolved by #348

**Comment by AndrewBarba** [comment]
@hunterloftis This complete broke existing apps that had a yarn file.

**Comment by juanpujol** [comment]
Yes, apps with the env var `NPM_CONFIG_PRODUCTION: false` broke for me because Yarn won't recognize that and install devDependencies. I went back to NPM for now. 

Anyone else had the same issue?

**Comment by AndrewBarba** [comment]
@juanpujol We don't set that flag, don't want any dev dependencies, and Yarn still broke for us. Our package.json only has 5 production deps on the app I posted above

**Comment by hunterloftis** [comment]
@AndrewBarba, Heroku isn't doing anything special with yarn. Just `yarn install --pure-lockfile --ignore-engines`. So if it's really working locally, you shouldn't have any issues. The things to check are:

- what versions of node and yarn are you using locally, and do they match the versions you've specified in package.json/engines?
- can you reproduce building your app from scratch (starting with `rm -rf node_modules` and a clean yarn cache) locally?

**Comment by joeyespo** [comment]
@juanpujol I did about a month ago. ([See my comment above](https://github.com/heroku/heroku-buildpack-nodejs/issues/337#issuecomment-265850065).)

Just waiting on a new Yarn release that includes https://github.com/yarnpkg/yarn/issues/1975 now.

**Comment by juanpujol** [comment]
Thank you @joeyespo

**Comment by ecozoic** [comment]
Yah this seem to be busted. I don't need any devDependencies for my apps and I'm getting all sorts of errors when using yarn.

**Comment by AndrewBarba** [comment]
@hunterloftis I did. Node version matches. Wiped everything, `yarn`, works great locally. Yarn version is `0.17.10`

**Comment by hunterloftis** [comment]
@ecozoic @AndrewBarba I'd recommend opening tickets so I can help you with specific apps. Several thousand apps have already built with this in place without issues, but it could be that your particular dependency trees hit a yarn bug (of which there are plenty at the moment).

**Comment by adamreisnz** [comment]
For what it's worth, my app is building and deploying fine with the yarn process in place it seems, so the issues that are being reported are probably specific to certain configurations/packages.

**Comment by AndrewBarba** [comment]
This is my package.json, not exactly anything crazy:

```
{
  "name": "...",
  "description": "...",
  "author": "...",
  "private": true,
  "engines": {
    "node": "6.9.1"
  },
  "scripts": {
    "start": "node server"
  },
  "dependencies": {
    "bluebird": "3.4.6",
    "compression": "1.6.2",
    "express": "4.14.0",
    "forky": "0.3.0",
    "serve-static": "1.11.1"
  }
}
```

**Comment by natecox** [comment]
@AndrewBarba I think I missed something here. Isn't the issue that you don't want dev dependencies installed on production? From your `package.json` it doesn't look like any of your dependencies are marked as dev only.

**Comment by AndrewBarba** [comment]
@natecox Someone else made that comment, in a previous comment I said that we didn't have any dev dependencies and it is still breaking for us. First person to mention there was an issue said that a config var was being ignored, but that is an open issue in Yarn. For me, i Just want production dependencies, and on a clean installation of yarn, no node modules folder, running `yarn` works correctly locally. It doesn't on Heroku

**Comment by natecox** [comment]
Ah, got it.

**Comment by hunterloftis** [comment]
@AndrewBarba yarn relies on a lockfile. What does yours look like?

**Comment by AndrewBarba** [comment]
@hunterloftis Yes, I'm aware. Thats why this is so concerning. The whole point of Yarn is to install exactly what is in the lock file. The same one checked into source, on all dev machines, and Heroku. When everything runs locally, all CI passes, Heroku build passes, server goes to start and THAT's where it breaks down... not fun. If build breaks fine. But thinking the install was successful and deploying to all nodes? Terrifying.

Here is lock: [https://assets.tablelist.com/fyre/yarn.lock.zip](https://assets.tablelist.com/fyre/yarn.lock.zip)

**Comment by AndrewBarba** [comment]
Okay digging into actual build logs, this is what a successful looks like (was able to get this to succeed just now on a dev app with NODE_ENV=development):

```
-----> Building dependencies
       Installing node modules (yarn)
       yarn install v0.17.10
       [1/4] Resolving packages...
       warning gulp > vinyl-fs > glob-stream > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
       warning gulp > vinyl-fs > glob-watcher > gaze > globule > minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
       warning gulp > vinyl-fs > glob-watcher > gaze > globule > glob > graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.
       success Already up-to-date.
       Done in 3.09s.
-----> Caching build
       Clearing previous node cache
       Saving 2 cacheDirectories (default):
       - node_modules
       - bower_components (nothing to cache)
-----> Build succeeded!
...
├─ async@0.2.10
       ├─ asynckit@0.4.0
       ├─ autoprefixer@6.5.4
       │  ├─ browserslist@~1.4.0
       │  ├─ caniuse-db@^1.0.30000597
       │  ├─ normalize-range@^0.1.2
       │  ├─ num2fraction@^1.2.2
       │  ├─ postcss-value-parser@^3.2.3
       │  └─ postcss@^5.2.6
```

And this is what the build looked like when the app couldn't start:

```
-----> Building dependencies
       Installing node modules (yarn)
       yarn install v0.17.10
       [1/4] Resolving packages...
       success Already up-to-date.
       Done in 0.57s.
-----> Caching build
       Clearing previous node cache
       Saving 2 cacheDirectories (default):
       - node_modules
       - bower_components (nothing to cache)
-----> Build succeeded!
...
?????? async@0.2.10
       ?????? asynckit@0.4.0
       ?????? autoprefixer@6.5.4
       ???  ?????? browserslist@~1.4.0
       ???  ?????? caniuse-db@^1.0.30000597
       ???  ?????? normalize-range@^0.1.2
       ???  ?????? num2fraction@^1.2.2
       ???  ?????? postcss-value-parser@^3.2.3
       ???  ?????? postcss@^5.2.6
```

So something strange going on between the two builds. Almost like it thought that it cached deps on my production app

**Comment by AndrewBarba** [comment]
I was just able to deploy to my production app with `NODE_MODULES_CACHE=false`, same yarn.lock file as before

**Comment by adamreisnz** [comment]
I guess I spoke to soon. Getting `Error: Cannot find module '/app/node_modules/bcrypt/lib/binding/bcrypt_lib.node'` error now in staging environment. Will try to disable `NODE_MODULES_CACHE` and see if that helps.

**Comment by adamreisnz** [comment]
I think in light of all the issues people are experiencing, maybe the build pack should not *automatically* use `yarn` if there is a `yarn.lock` file present, but rather work off what is specified in package.json. Maybe if `npm` is specified as an engine, use `npm`. If `yarn` is, then use `yarn`. Or some other Heroku specific setting..

**Comment by gordonk** [comment]
@hunterloftis I was unable to deploy my node/yarn application and setting the build pack back to pre-yarn was only solution mentioned here that worked. Otherwise crashed when trying to start the express server.

```
2016-12-17T08:35:17.436016+00:00 app[web.1]: Error: Cannot find module 'mime'
2016-12-17T08:35:17.436017+00:00 app[web.1]:     at Function.Module._resolveFilename (module.js:469:15)
```

Npm install worked fine.

**Comment by adamreisnz** [comment]
We just had to remove the `yarn.lock` file from our project because the build was failing in production 😢 

I guess we'll wait a bit longer until yarn is more stable.

**Comment by drewhamlett** [comment]
Are you using the master branch?  We were using yarn branch but it seems the master branch supports yarn now.  It's been working for us.

**Comment by adamreisnz** [comment]
Yes, master branch

**Comment by pasupulaphani** [comment]
yeah and it does respect yarn version put under engines

```
  "engines": {
    "node": "7.0.x",
    "yarn": "0.16.1"
  },
```

**Comment by tribou** [comment]
For anyone else searching, I found a couple more workarounds until this issue gets fixed:

- Add `yarn.lock` to a [`.slugignore` file](https://devcenter.heroku.com/articles/slug-compiler#ignoring-files-with-slugignore) so Heroku goes back to using NPM for now. (solution I chose)

OR

- Set your buildpack to the [last working version](https://devcenter.heroku.com/articles/nodejs-support#build-behavior) with `heroku buildpacks:set https://github.com/heroku/heroku-buildpack-nodejs#v91`

**Comment by robinvdvleuten** [comment]
Same here! Build is failing because the app requires any private modules...maybe a warning ahead by Heroku was in place here. A relative new tool like Yarn still has a lot of rough edges, maybe wait for a couple of months to merge such thing in a production platform or make it a beta feature for users who want to run their app on the wild side of things.

**Comment by joshmanders** [comment]
@robinvdvleuten you can opt out of using yarn by adding `yarn.lock` to your `.slugignore` file. Putting Yarn into this buildpack is one of the best ways to help the Yarn team find all edge cases and get them fixed, because they have a higher sample of users to get info from.

**Comment by freeall** [comment]
I'm guessing Heroku didn't know that yarn doesn't support private github modules - which quite a lot of people use

**Comment by hunterloftis** [comment]
@freeall @robinvdvleuten  yes, we're aware that yarn doesn't support private modules. Why are you using yarn / have a yarn.lock file in the root of your project, if you are relying on functionality that yarn doesn't yet support? Yarn is only used in the buildpack if you check `yarn.lock` into git.

Yarn is a new technology, and while it's building the vast majority of apps faster and with determinism, it may not be suited for your project. If you still want to have a `yarn.lock` file, but simultaneously do not want to use yarn, I've added instructions for opting out here:

- https://devcenter.heroku.com/articles/nodejs-support#build-behavior

**Comment by hunterloftis** [comment]
Yarn updates:

https://github.com/heroku/heroku-buildpack-nodejs/issues/359

**Comment by joeyespo** [comment]
Looks like there's another problem when a `devDependencies` includes binaries (when `NODE_ENV=production` and `NPM_CONFIG_PRODUCTION=false`) https://github.com/yarnpkg/yarn/issues/2214

**Comment by nascVi** [comment]
Can't say that much about Yarn! Thanks landscapersParis is also on slack as known as geeks gardens and landscapers

**Comment by fozzle** [comment]
It seems to me that anytime I use the default cache settings any packages with native bindings will fail after the first deploy. Disabling the module cache fixes this (as many have noted) but I'm curious if anyone has figured out a way to make the cache play nicely with native bindings now that it's been a few months.
